### ISSUE TITLE
Automatic casting of nullable nested array DTOs

### ISSUE THREAD

**Issue by pikant** [spatie/data-transfer-object#167]
I work with the Shopify API, which has the special feature that you can use a `fields` parameter with every GET request to specify which fields should be included in the response. This procedure is intended to save data volume.

This results in the specialty that I have to make all fields of my DTOs nullable. This also works so far, only with nested arrays (see https://github.com/spatie/data-transfer-object#automatic-casting-of-nested-array-dtos) it doesn't seem to work.

I am getting the following error:
```
   Spatie\DataTransferObject\DataTransferObjectError 

  Invalid type: expected `App\Domain\Shopify\DataTransferObjects\Order\Order::line_items` to be of type `\App\Domain\Shopify\DataTransferObjects\Order\LineItems[]`, instead got value `null`, which is NULL..

  at vendor/spatie/data-transfer-object/src/DataTransferObjectError.php:24
     20▕         $msg = count($invalidTypes) > 1
     21▕             ? "The following invalid types were encountered:\n" . implode("\n", $invalidTypes) . "\n"
     22▕             : "Invalid type: {$invalidTypes[0]}.";
     23▕ 
  ➜  24▕         throw new self($msg);
     25▕     }
     26▕ 
     27▕     public static function invalidTypeMessage(
     28▕         string $class,
```

Here is an excerpt from the DTO:
```php
class Order extends FlexibleDataTransferObject
{
    public? int $ id;

    // ...

    / ** @var \ App \ Domain \ Shopify \ DataTransferObjects \ Order \ LineItems [] $ line_items * /
    public? array $ line_items;
```

I'm about to create a PR with some tests that reproduces the behavior. I will link this PR here.

**Comment by mfn** [comment]
Is it a problem with contradicting phpdoc? phpdoc does not say it's nullable and I think it has higher precedence if present?

**Comment by pikant** [comment]
Thanks for the quick answer. Indeed I tried a trailing ' | null' in the phpdoc. However, I just realized, that it only works without the whitespaces. Is this a phpdoc constraint?

**Comment by pikant** [comment]
I also updated this change in my PR and renamed it also. Only in case you want to include the tests in the code: https://github.com/spatie/data-transfer-object/pull/168

This issue here can be closed.

Many Thanks!

**Comment by mfn** [comment]
> However, I just realized, that it only works without the whitespaces

Never ever have I see these whitespaces phpdoc in all the years. I actually thought this was a freak copy paste mistake on your part:
```
/ ** @var \ App \ Domain \ Shopify \ DataTransferObjects \ Order \ LineItems [] $ line_items * /
```
So yes, I would not expected this to work. Especially since the "magic marker" has to be `/**` (almost "IMHO definitely") to work etc.

But yes, I see that the "real" `use` statement does not care about whitespaces, probably also due to "namespace grouping" and stuff.

TL;DR: `/** @var \App\Domain\Shopify\DataTransferObjects\Order\LineItems[]|null $line_items */` it would be in my book

**Comment by pikant** [comment]
Yes, that was an unfortunate coincidence: My text editor created this inappropriate whitespaced. I hadn't focused that much on this code snippet as I actually provided a PR demonstrating the behavior. However, I actually had a white space in front of the pipe (` | null`) that crashed it.

The error message that comes up when working with the one whitespace in around the pipe confused me very much:
```
1) Spatie\DataTransferObject\Tests\NestedNullableArrayFlexibleDataTransferObjectTest::it_can_validate_nested_arrays
Spatie\DataTransferObject\DataTransferObjectError: Invalid type: expected `Spatie\DataTransferObject\Tests\FlexibleWrapperWithNullableArrayDtoPropertyType::arr` to be of type `\Spatie\DataTransferObject\Tests\NestedArrayDtoPropertyType[] ,  null`, instead got value `array`..
```

It seemed to me that the library had already understood that the type could also be null:
```
expected ... to be of type `\Spatie\DataTransferObject\Tests\NestedArrayDtoPropertyType[] ,  null
```

But thank you for your support anyway!
